kimchi/src/folding/instance_witness.rs

```diff
@@ -0,0 +1,170 @@
+    fn relax(
+        self,
+        zero_poly: &[G::ScalarField],
+        zero_commit: PolyComm<G>,
+    ) -> RelaxedInstance<G, Self>;
+}
+
+pub struct RelaxedWitness<G: CommitmentCurve, W: WitnessTrait<G>> {
+    pub witness: W,
+    pub error_vec: Vec<G::ScalarField>,
+}
+
+pub trait RelaxableInstance<G: CommitmentCurve, I: InstanceTrait<G>> {
+    fn relax(self, zero_commitment: PolyComm<G>) -> RelaxedInstance<G, I>;
+}
+impl<G: CommitmentCurve, I: InstanceTrait<G>> RelaxableInstance<G, I> for I {
+    fn relax(self, zero_commitment: PolyComm<G>) -> RelaxedInstance<G, I> {
+        self.relax(zero_commitment)
+    }
+}
+impl<G: CommitmentCurve, I: InstanceTrait<G>> RelaxableInstance<G, I> for RelaxedInstance<G, I> {
+    fn relax(self, _zero_commitment: PolyComm<G>) -> RelaxedInstance<G, I> {
+        self
+    }
+}
+pub trait RelaxableWitness<G: CommitmentCurve, W: WitnessTrait<G>> {
+    fn relax(self, zero_poly: &[G::ScalarField]) -> RelaxedWitness<G, W>;
+}
+impl<G: CommitmentCurve, W: WitnessTrait<G>> RelaxableWitness<G, W> for W {
+    fn relax(self, zero_poly: &[G::ScalarField]) -> RelaxedWitness<G, W> {
+        self.relax(zero_poly)
+    }
+}
+impl<G: CommitmentCurve, W: WitnessTrait<G>> RelaxableWitness<G, W> for RelaxedWitness<G, W> {
+    fn relax(self, _zero_poly: &[G::ScalarField]) -> RelaxedWitness<G, W> {
+        self
+    }
+}
+
+// pub trait RelaxableInstance<G: KimchiCurve, >: seal::Seal<G, I, 0> {
+pub trait RelaxablePair<G: CommitmentCurve, I: InstanceTrait<G>, W: WitnessTrait<G>> {
+    fn relax(
+        self,
+        zero_poly: &[G::ScalarField],
+        zero_commitment: PolyComm<G>,
+    ) -> (RelaxedInstance<G, I>, RelaxedWitness<G, W>);
+}
+impl<G, I, W> RelaxablePair<G, I, W> for (I, W)
+where
+    G: CommitmentCurve,
+    I: InstanceTrait<G> + RelaxableInstance<G, I>,
+    W: WitnessTrait<G> + RelaxableWitness<G, W>,
+{
+    fn relax(
+        self,
+        zero_poly: &[G::ScalarField],
+        zero_commitment: PolyComm<G>,
+    ) -> (RelaxedInstance<G, I>, RelaxedWitness<G, W>) {
+        let (instance, witness) = self;
+        (
+            RelaxableInstance::relax(instance, zero_commitment),
+            RelaxableWitness::relax(witness, zero_poly),
+        )
+    }
+}
+
+trait Pair2<G: CommitmentCurve> {
+    type Instance: InstanceTrait<G>;
+    type Witness: WitnessTrait<G>;
+}
+
+impl<G: CommitmentCurve, I: InstanceTrait<G>> RelaxedInstance<G, I> {
+    pub fn add_error(self, error_commitment: &PolyComm<G>, challenge: G::ScalarField) -> Self {
+        let RelaxedInstance {
+            instance,
+            u,
+            error_commitment: error,
+        } = self;
+        let error_commitment = &error - &error_commitment.scale(challenge);
+        RelaxedInstance {
+            instance,
+            u,
+            error_commitment,
+        }
+    }
+
+    pub(super) fn combine(a: Self, b: Self, challenge: <G>::ScalarField) -> Self {
```

Comment:
maybe using the same name was not the best idea, the struct level methods are extensions based on the method of the underlying type implementing the trait.
I just need `RelaxedInstance` and the like to be opaque types and use the methods internally.

Reply:
I think one thing I do not find intuitive in this hierarchy of structs/traits is that `InstanceTrait` and `WitnessTrait` have the `relax` function default implementation. Not all the instances/witnesses are relaxable, and you have a specific trait for those which are: `RelaxedInstance`/`RelaxedWitness`. Maybe a better way would be to use the `relax` from them.

It's a nit perhaps but this hierarchy is quite confusing.

Reply:
actually, all instances/pairs are relaxable, the ones already relaxed have a trivial implementation that is just the identity function.
And there is no specific traits for `RelaxedInstance`/`RelaxedWitness`, just type-level methods.
I see it as similar to taking `Into<T>` instead of `T` as the argument in some function for example.